MaxText/layers/moe.py

```diff
@@ -403,71 +493,54 @@ def gmm(inputs, kernel, group_sizes):
+            scales = jax.lax.pad(scales, jnp.array(0.0, dtype=scales.dtype), [(0, pad_length, 0), (0, 0, 0)])
+          output *= scales
+      if hs_shape[0] % PAD_LENGTH:
         output = output[: hs_shape[0]]
       return output
 
-    def get_all_to_all_params(all_shards_group_sizes, local_expert_size, num_expert_shards):
-
-      class TransformStrategy(Enum):
-        INPUT_OFFSET = auto()
-        SEND_SIZE = auto()
-        OUTPUT_OFFSET = auto()
-        RECV_SIZE = auto()
-
-      def transform_array(input_array, shard_id, strategy):
-        """This function transforms the input array based on the specified strategy,
-        preparing it for the usage with `ragged_all_to_all` API. The transformation
-        determines how data is sent and received between shards.
-        """
-        if strategy == TransformStrategy.INPUT_OFFSET:
-          # Index of input array for the send
-          local_array = input_array[shard_id]
-          return jnp.concatenate((jnp.array([0]), jnp.cumsum(local_array)[:-1]))
-        elif strategy == TransformStrategy.SEND_SIZE:
-          # Size of input array for the send
-          return input_array[shard_id]
-        elif strategy == TransformStrategy.OUTPUT_OFFSET:
-          # Received index in the target output
-          zero_row = jnp.zeros((1,) + input_array.shape[1:], dtype=input_array.dtype)
-          array_with_zeros = jnp.concatenate((zero_row, input_array), axis=0)
-          cumulated_array = jnp.cumsum(array_with_zeros, axis=0, dtype=input_array.dtype)
-          return cumulated_array[shard_id]
-        elif strategy == TransformStrategy.RECV_SIZE:
-          # Received size in the traget output
-          return input_array[:, shard_id]
-        else:
-          raise ValueError(f"Unknown tranform array strategy: {strategy}")
-
-      local_id = jax.lax.axis_index("expert")
-      input_offsets = transform_array(all_shards_group_sizes, local_id, TransformStrategy.INPUT_OFFSET)
-      send_sizes = transform_array(all_shards_group_sizes, local_id, TransformStrategy.SEND_SIZE)
-      output_offsets = transform_array(all_shards_group_sizes, local_id, TransformStrategy.OUTPUT_OFFSET)
-      recv_sizes = transform_array(all_shards_group_sizes, local_id, TransformStrategy.RECV_SIZE)
-      return input_offsets, send_sizes, output_offsets, recv_sizes
-
     # Currently, we only support data and tensor parallelism with Megablox.
     # We all gather the input activations over tensor parallelism to follow strategy
     # in https://parsa.epfl.ch/course-info/cs723/papers/Megatron.pdf.
-    input_partition_pspec = nn.logical_to_mesh_axes(("activation_batch", None, None))
-    gate_logits_pspec = nn.logical_to_mesh_axes(("activation_batch", None, None))
+    
+    # Check if the batch should be sharded by expert and whether the batch_size supports this.
+    # Useful for Interleaved Inference where Prefill always has batch_size=1 while Decode
+    # can have batch_size > 1.
+    try:
```

Comment:
Thinking if we could simplify this logic a little bit. No better idea yet, let me think about it.